### PR TITLE
[ADD] py.js: len() and bool() on tuple, list and dict types

### PR DESCRIPTION
**Pull request by Whisno**
and use it to set a color on bank.statement.line in list views, to indicate if it was reconciled.

### CODE REVIEW

File: addons/account/account_view.xml [hunk 1]
```diff
@@ -600,7 +600,7 @@
                     <notebook>
                         <page string="Transactions" name="statement_line_ids">
                             <field name="line_ids" context="{'date':date}">
-                                <tree editable="bottom" string="Statement lines">
+                                <tree editable="bottom" string="Statement lines" colors="grey:bool(journal_entry_ids)==True">
```
Comment (xmo-odoo) [inline addons/account/account_view.xml]: Why not just `bool(journal_entry_ids)`?

File: addons/web/static/lib/py.js/lib/py.js [hunk 2]
```diff
@@ -969,6 +969,15 @@ var py = {};
         __init__: function () {
             this._values = [];
         },
+        __len__: function () {
+            return this._values.length;
+        },
+        __nonzero__: function () {
+            var len = this.__len__();
+            if (len === 0) return py.False;
+            if (len === 1 && py.PY_not(this._values[0])) return py.False;
```
Comment (xmo-odoo) [inline addons/web/static/lib/py.js/lib/py.js]: That's not correct. Tuple and list have the same truthiness rules.

File: addons/web/static/lib/py.js/lib/py.js [hunk 3]
```diff
@@ -969,6 +969,15 @@ var py = {};
         __init__: function () {
             this._values = [];
         },
+        __len__: function () {
+            return this._values.length;
+        },
+        __nonzero__: function () {
+            var len = this.__len__();
```
Comment (xmo-odoo) [inline addons/web/static/lib/py.js/lib/py.js]: Should use `py.PY_size(this)`

File: addons/web/static/lib/py.js/lib/py.js [hunk 4]
```diff
@@ -1013,6 +1027,12 @@ var py = {};
         __setitem__: function (key, value) {
             this._store[key.__hash__()] = [key, value];
         },
+        __len__: function () {
+            return Object.keys(this._store).length
+        },
+        __nonzero__: function () {
+            return this.__len__ > 0;
```
Comment (xmo-odoo) [inline addons/web/static/lib/py.js/lib/py.js]: Doesn't call the method, should use `py.PY_size(this)`.

**Comment by xmo-odoo** [comment]
Fixing the prefix-not case should be simple enough: line 323 should simply check that `tokens` is non-empty before trying to check the `id` of the last token in the array. That would allow `blue:not journal_entry_ids` (or at least `blue: not bool(journal_entry_ids)`)

**Comment by xmo-odoo** [comment]
And a slightly better improvement might be for view_list to call `py.PY_isTrue` on the result of `py.evaluate`, that way the explicit conversion is unnecessary and the actual result matches expectations (of Pythonic truthiness)

File: addons/web/static/lib/py.js/lib/py.js (hunk 2 above)
Comment (Whisno) [inline addons/web/static/lib/py.js/lib/py.js]: Not in my python shell (altough maybe for some obscure immutability related reason, I don't understand why it wouldn't) : 
```
17:54 $ python --version
Python 2.7.8
17:54 $ python
Python 2.7.8 (default, Oct 20 2014, 15:05:19) 
[GCC 4.9.1] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> bool([False])
True
>>> bool((False))
False
```

File: addons/account/account_view.xml (hunk 1 above)
Comment (Whisno) [inline addons/account/account_view.xml]: Ok, done

File: addons/web/static/lib/py.js/lib/py.js (hunk 3 above)
Comment (Whisno) [inline addons/web/static/lib/py.js/lib/py.js]: Fixed

File: addons/web/static/lib/py.js/lib/py.js (hunk 4 above)
Comment (Whisno) [inline addons/web/static/lib/py.js/lib/py.js]: Oops, my bad

File: addons/web/static/lib/py.js/lib/py.js (hunk 2 above)
Comment (guewen) [inline addons/web/static/lib/py.js/lib/py.js]:
```
>>> bool((False))
```
Is not a tuple, only redundant parentheses.
```
>>> bool((False,))
```
Is a tuple and returns True

Comment (Whisno) [inline addons/web/static/lib/py.js/lib/py.js]: Oops, the python newb that I am needs to RTFM rather soon. Thanks for the clarification

**Pull request closed without merging**